ke/compiler.py

```diff
@@ -214,6 +214,20 @@ def compile_either(e, macros):
             elif isinstance(c, asm.CharacterClass):
                 characters += c.characters
         return asm.CharacterClass(characters, False)
+    
+    non_empty_elements = 0
+    non_empty = None
```

Comment:
I feel like this can be simplified. How about something like:

```
all_nonempty = filter(is_not_empty, compiled)
if len(all_nonempty) == 1:
    is_greedy = is_nonempty(compiled[0])
    (subexpr,) = all_nonempty
    return asm.Multiple(0, 1, is_greedy, subexpr)
```

I think it's easier to understand what the rules are meant to be from this, right?

Reply:
No filter for all_nonempty. Also did you mean is_not_empty instead of is_nonempty ?